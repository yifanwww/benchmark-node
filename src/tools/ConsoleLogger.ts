import chalk from 'chalk';

export enum LogKind {
    Default,
    Header,
    Statistic,
    Info,
    Error,
}

type ColorScheme = {
    [key in LogKind]: (value: string) => string;
};

export class ConsoleLogger {
    public static readonly default: ConsoleLogger = new ConsoleLogger();

    private colorScheme: ColorScheme;

    public constructor(colorScheme?: ColorScheme) {
        this.colorScheme = colorScheme ?? this.createColorfulScheme();
    }

    private createColorfulScheme = (): ColorScheme => ({
        [LogKind.Default]: chalk.white,
        [LogKind.Header]: chalk.magentaBright,
        [LogKind.Statistic]: chalk.cyanBright,
        [LogKind.Info]: chalk.yellow,
        [LogKind.Error]: chalk.red,
    });

    private _writeImpl = (text: string) => process.stdout.write(text);
    private _writeLineImpl = (text?: string) => process.stdout.write(text ? `${text}\n` : '\n');

    public write(text: string): void;
    public write(kind: LogKind, text: string): void;

    public write(...args: [string] | [LogKind, string]) {
        if (args.length === 1) {
            this._write(LogKind.Default, this._writeImpl, args[0]);
        } else {
            this._write(args[0], this._writeImpl, args[1]);
        }
    }

    public writeLine(): void;
    public writeLine(text: string): void;
    public writeLine(kind: LogKind, text: string): void;

    public writeLine(...args: [] | [string] | [LogKind, string]) {
        if (args.length === 0) {
            this._writeLineImpl();
        } else if (args.length === 1) {
            this._write(LogKind.Default, this._writeLineImpl, args[0]);
        } else {
            this._write(args[0], this._writeLineImpl, args[1]);
        }
    }

    private _write(kind: LogKind, write: (text: string) => void, text: string) {
        write(this.colorScheme[kind](text));
    }

    public writeHeader = (text: string) => this.write(LogKind.Header, text);
    public writeStatistic = (text: string) => this.write(LogKind.Statistic, text);
    public writeInfo = (text: string) => this.write(LogKind.Info, text);
    public writeError = (text: string) => this.write(LogKind.Error, text);

    public writeLineHeader = (text: string) => this.writeLine(LogKind.Header, text);
    public writeLineStatistic = (text: string) => this.writeLine(LogKind.Statistic, text);
    public writeLineInfo = (text: string) => this.writeLine(LogKind.Info, text);
    public writeLineError = (text: string) => this.writeLine(LogKind.Error, text);
}
