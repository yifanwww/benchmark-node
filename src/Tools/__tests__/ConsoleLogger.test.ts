import { ConsoleLogger } from '../ConsoleLogger';

describe(`Test class \`${ConsoleLogger.name}\``, () => {
    beforeEach(() => {
        jest.spyOn(process.stdout, 'write').mockImplementation((str) => {
            expect(str).toMatchSnapshot();
            return true;
        });
    });

    it(`calls method ${ConsoleLogger.prototype.write.name}`, () => {
        expect(ConsoleLogger.default.write('Hello world'));
    });

    it('calls method writeHeader', () => {
        expect(ConsoleLogger.default.writeHeader('Hello world'));
    });

    it('calls method writeStatistic', () => {
        expect(ConsoleLogger.default.writeStatistic('Hello world'));
    });

    it('calls method writeInfo', () => {
        expect(ConsoleLogger.default.writeInfo('Hello world'));
    });

    it('calls method writeError', () => {
        expect(ConsoleLogger.default.writeError('Hello world'));
    });

    it(`calls method ${ConsoleLogger.prototype.writeLine.name}`, () => {
        expect(ConsoleLogger.default.writeLine());
        expect(ConsoleLogger.default.writeLine('Hello world'));
    });

    it('calls method writeLineHeader', () => {
        expect(ConsoleLogger.default.writeLineHeader('Hello world'));
    });

    it('calls method writeLineStatistic', () => {
        expect(ConsoleLogger.default.writeLineStatistic('Hello world'));
    });

    it('calls method writeLineInfo', () => {
        expect(ConsoleLogger.default.writeLineInfo('Hello world'));
    });

    it('calls method writeLineError', () => {
        expect(ConsoleLogger.default.writeLineError('Hello world'));
    });
});
