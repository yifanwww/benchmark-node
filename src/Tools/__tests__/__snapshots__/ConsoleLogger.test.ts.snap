// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test class \`ConsoleLogger\` calls method write 1`] = `"[37mHello world[39m"`;

exports[`Test class \`ConsoleLogger\` calls method writeError 1`] = `"[31mHello world[39m"`;

exports[`Test class \`ConsoleLogger\` calls method writeHeader 1`] = `"[95mHello world[39m"`;

exports[`Test class \`ConsoleLogger\` calls method writeInfo 1`] = `"[33mHello world[39m"`;

exports[`Test class \`ConsoleLogger\` calls method writeLine 1`] = `
"
"
`;

exports[`Test class \`ConsoleLogger\` calls method writeLine 2`] = `
"[37mHello world[39m
"
`;

exports[`Test class \`ConsoleLogger\` calls method writeLineError 1`] = `
"[31mHello world[39m
"
`;

exports[`Test class \`ConsoleLogger\` calls method writeLineHeader 1`] = `
"[95mHello world[39m
"
`;

exports[`Test class \`ConsoleLogger\` calls method writeLineInfo 1`] = `
"[33mHello world[39m
"
`;

exports[`Test class \`ConsoleLogger\` calls method writeLineStatistic 1`] = `
"[96mHello world[39m
"
`;

exports[`Test class \`ConsoleLogger\` calls method writeStatistic 1`] = `"[96mHello world[39m"`;
